fix: update consensus block for requests between state update ticks

mariuspod opened the pull request:
<!-- Contributions welcome! See https://github.com/ethereum-optimism/.github/blob/master/CONTRIBUTING.md -->

**Description**

Hey, we observed a 20-25% error rate on proxyd loadbalancing multiple backends,

During rewriting request params, there's an exception returned, if the incoming request arrives just in between the consensus state update ticks. With the default proxyd intervals, this might happen sub-second in normal mode or sub-2-second in HA mode. In our tests, the `RewriteContext` contained an outdated latest block at the time the request was processed although consensus already had a newer block available but as the `rctx.latest` field was populated a few ms before, the handler crashed on the incoming request.

I've fixed the error by using a simple retry logic which polls the latest consensus block from the consensus tracker.
In most cases this should not add much to the overall latencies but it should return a correct result to the caller. With this fix applied, our error rates dropped to zero :partying_face: 

I've also added a new config flag `consensus_poller_retry` to toggle this new behaviour.

Also created 3 new metrics for consensus related variables which helped me a lot to debug this. Can be also removed but they were quite helpful :sweat_smile: 

Tested on stacks with consensus HA and normal mode.

Please let me know what you think or how this could be improved.

**Tests**

Added 3 more tests to the `rewriter_test.go`

mariuspod committed:
feat: handling edge case for proxyd consensus poller during rewriting request params

mariuspod committed:
feat: extract retry logic to separate method and also use for rewriteTagBlockNumberOrHash()

mariuspod committed:
fix: typo

trianglesphere commented on proxyd/rewriter.go:
```diff
@@ -301,10 +305,39 @@ func rewriteTagBlockNumberOrHash(rctx RewriteContext, current *rpc.BlockNumberOr
+		// check if consensus has a newer block already
+		// increase the sleep time in each iteration by 10ms
+		// this will sleep at most 2100ms in total (20*21/2*10)
+		for i := 1; i < 21; i++ {
```
should use the retry package in op-service

mariuspod replied:
done

trianglesphere reviewed (commented):
cc @felipe-op if we want to bring these changes in

mariuspod committed:
feat: tracking more metrics to see how much latencies are added

mariuspod committed:
feat: always track the metrics for retries and totalSleepMs

mariuspod committed:
feat: update consensus by reading the redis state and populating in-memory state from it

mariuspod committed:
feat: use retry package as suggested during pr review

mariuspod commented:
We fixed another issue with added latencies due to outdated in-memory consensus state.
There are two new metrics:
1. number of retries
2. total delays during retries

As the in-memory state only updates during heartbeat ticks, the latencies can become quite big.

So in the error case, instead of the simple polling, we're now actively querying the redis backend to retrieve the current consensus state so it's updated in-memory with the latest state. In the screenshots you can see the additional delays and retry counts before and after the new fix (blue: max, purple: avg)
![consensus_updates](https://github.com/ethereum-optimism/optimism/assets/14898268/dc09fe37-d403-4d86-a721-d4e2f00bc028)

As suggested, I modified the retry logic to use the module from `optimism-service`, this is still useful for non-HA mode IMO but is optional if you prefer to remove that code.

mariuspod committed:
feat: replace boolean flag with int flag to specify max retries when there is an outdated consensus context

mariuspod commented:
refactored and renamed config flag from boolean to int
`consensus_max_retries` default: 0

The new metrics are:
- `group_consensus_candidate_lag`
- `group_consensus_requested_block`
- `group_consensus_current_consensus_block`
-  `group_consensus_block_update_retries`
-  `group_consensus_block_update_total_sleep_ms`

mariuspod commented:
rebased

mariuspod commented:
@trianglesphere we're running this fix in prod since a few weeks.
Without it, we're getting many errors due to client requests being loadbalanced between multiple proxyd instances for subsequent requests:

`{"code":-32019,"message":"block is out of range"}`

trianglesphere commented:
@mariuspod you should ping @felipe-op

mariuspod commented:
hey @felipe-op
is there anything I could add to this pr ?

felipe-op commented:
Thanks for tagging. 

For now, I just want to make you aware proxyd is moving here: https://github.com/ethereum-optimism/infra

I appreciate if you can open the PR in the infra repo.
